src/savegame.cpp

```diff
@@ -635,69 +643,1183 @@ void ingame_savegame::write_game(config_writer &out) {
+		suffix = "_1_12";
+
+	return suffix;
+}
+
+std::string append_suffix_to_unit_type(std::string unit_type, config& cfg)
+{
+	const std::string suffix = convert_version_to_suffix(cfg);
+	return unit_type + suffix;
+}
+
+void update_start_unit_types(config& replay_start, config& cfg)
+{
+	const std::string suffix = convert_version_to_suffix(cfg);
+	if(not suffix.empty())
+	{
+		for(config& side : replay_start.child_range("side"))
+		{
+ 			std::ostringstream ss, tt, uu;
+ 			// BFW 1.4, 1.6
+ 			std::string sep = "";
+			for(const std::string& leader : utils::split(side["leader"], ','))
+			{
+				ss << sep << leader << suffix;
+				sep = ",";
+			}
+			std::string new_leader = ss.str();
+			side["leader"] = new_leader;
+ 			if (side.has_attribute("recruit"))
+			{
+ 				sep = "";
+				for(const std::string& recruit : utils::split(side["recruit"], ','))
+				{
+					uu << sep << recruit << suffix;
+					sep = ",";
+				}
+				std::string new_recruit = uu.str();
+				side["recruit"] = new_recruit;
+			}
+ 			if (side.has_attribute("random_leader"))
+			{
+ 				sep = "";
+				for(const std::string& leader : utils::split(side["random_leader"], ','))
+				{
+					tt << sep << leader << suffix;
+					sep = ",";
+				}
+				std::string new_random = tt.str();
+				side["random_leader"] = new_random;
+			}
+ 			if (side.has_attribute("type")) side["type"] = side["type"].str()+suffix;
+
+ 			// BFW 1.8
+ 			if (side.has_attribute("previous_recruits"))
+			{
+ 				sep = "";
+				for(const std::string& recruit : utils::split(side["previous_recruits"], ','))
+				{
+					uu << sep << recruit << suffix;
+					sep = ",";
+				}
+			std::string new_recruit = uu.str();
+			side["previous_recruits"] = new_recruit;
+			}
+			for(config& unit : side.child_range("unit"))
+			{
+				bool found = unit["type"].str().find(suffix, 0);
+				if(found == std::string::npos)
+				{
+					unit["type"] = unit["type"].str()+suffix;
+ 					std::ostringstream vv;
+ 					sep = "";
+					for(const std::string& leader : utils::split(unit["advances_to"], ','))
+					{
+						vv << sep << leader << suffix;
+						sep = ",";
+					}
+					std::string new_advance = vv.str();
+					unit["advances_to"] = new_advance;
+				}
+			}
+		}
+	}
+}
+
+void update_replay_unit_types(config& replay, config& cfg)
+{
+	const std::string suffix = convert_version_to_suffix(cfg);
+	if(suffix.empty()) return;
+	for(config& command : replay.child_range("command"))
+	{
+ 		// BFW 1.12
+ 		if (config& recruit = command.child("recruit"))
+		{
+ 			recruit["type"] = recruit["type"] + suffix;
+		}
+		if (config& attack = command.child("attack"))
+		{
+			if (attack.has_attribute("attacker_type"))
+			{
+				attack["attacker_type"] = attack["attacker_type"] + suffix;
+				attack["defender_type"] = attack["defender_type"] + suffix;
+			}
+		}
+	}
+}
+
+void feral_trait_1_8_enter_hex(config& event)
+{
+	std::stringstream ss;
+	ss << "		name=enter_hex\n";
```

Comment:
These sorts of passages don't really belong in the source. The WML in each such function should be extracted into a .cfg file which the function then loads by name.